pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -274,50 +277,67 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 		}
 	}
 
+	if klusterlet.DeletionTimestamp.IsZero() {
+		if !hasFinalizer(klusterlet, klusterletFinalizer) {
+			return n.addFinalizer(ctx, klusterlet, klusterletFinalizer)
+		}
+	}
+
 	// TODO: remove this when detached mode is not used in klusterlet
 	if config.InstallMode == operatorapiv1.InstallModeDetached {
 		config.InstallMode = operatorapiv1.InstallModeHosted
 	}
 
 	if config.InstallMode == operatorapiv1.InstallModeHosted {
+		var updated bool
+		var updateErr error
 		managedClusterClients, err = n.buildManagedClusterClientsHostedMode(ctx, n.kubeClient, config.AgentNamespace, config.ExternalManagedKubeConfigSecret)
 		if err != nil {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			_, updated, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionFalse, Reason: "KlusterletPrepareFailed",
 				Message: fmt.Sprintf("Failed to build managed cluster clients: %v", err),
 			}))
-			return err
 		} else {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			_, updated, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionTrue, Reason: "KlusterletPrepared",
 				Message: "Klusterlet is ready to apply",
 			}))
 		}
+		if updated {
+			return updateErr
+		}
+		// If the klusterletEven klusterletReadyToApply condition does not updated, even if building the managed cluster
+		// client fails, it will not return here, because it may need to execute the following code to do some cleanup
+		// work
 	}
 
-	// Update finalizer after the clients created, otherwise, for Hosted mode if users did not provide the
-	// external managed cluster kubeconfig, we can not delete the finalizer
-	if klusterlet.DeletionTimestamp.IsZero() {
-		hasFinalizer := false
-		for i := range klusterlet.Finalizers {
-			if klusterlet.Finalizers[i] == klusterletFinalizer {
-				hasFinalizer = true
-				break
-			}
+	// Klusterlet is deleting, we remove its related resources on managed and management cluster
+	if !klusterlet.DeletionTimestamp.IsZero() {
+		skip := skipCleanupManagedClusterResources(klusterlet, config.InstallMode)
+
+		if !skip && !readyToCleanupManagedClusterResources(klusterlet, config.InstallMode) {
```

Comment:
readyToCleanupManagedClusterResources seems not necessary, you just need to check the condition, right?